Allow *limited* post-compilation changes collision filter groups

SeanCurtis-TRI opened the pull request:
What *is* allowed:
  - Compiled bodies can be added to new collision filter groups.
  - New collision elements can be added to compiled bodies (and they
    inherit the previous collision filter groups and cliques)

What is *not* allowed:
  - Collision filter groups cannot be "removed"; that is, a body can only
    be added to new groups and not be removed from previously compiled
    groups. Similarly, groups can be added to its ignore set, but not
    removed.
  - Compiled filter groups *cannot* be referenced by name.

Implementation (what's in the commit)
  - Clean up
    - Unit tests use new `DRAKE_EXPECT_THROWS_MESSAGE`
    - common test functionality (creating bodies, performing collision test)
      refactored into utility functions.
    - Removed unnecessary namespace references.
    - General typo correct and formatting issues.
    - Removed no longer relevant method `SetBodyCollisionFilters()` from
      `RigidBodyTree` as per TODO.
    - Updated various doxygen documentation.
  - Feature support
    - Collision element's collision filter groups can be expanded using the
      new `merge_collision_filter()` method.
    - Collision model (model.h and bullet_model.{h|cc}) extended to include
      interface to report *changes* in collision filter state for a
      collision element.
    - `RigidBodyTree` updated to support new interface
      - No longer throws an exception when modifying compiled bodies.
      - `CompileCollisionState()` handles new collision data as well as
        previously compiled data.
    - New unit tests
      - Add element to compiled body.
      - Add collision filter group after compilation.
      - Confirm post-compile element inherits body's collision filter groups.
      - Confirm post-compile element inherits body's cliques.

Notes to reviewers:
 - In `rigid_body_tree.cc`, don't worry too much about the differences in `CompileCollisionState()` and `CreateCollisionCliques()`; they are basically rewrites.
 - Proved to be a larger-than-expected PR due to tests and documentation.
```
Category            added  modified  removed  
----------------------------------------------
code                495    73        68       
comments            217    59        35       
blank               63     0         1        
----------------------------------------------
TOTAL               775    132       104 
```

SeanCurtis-TRI commented:
+@rcory for feature review if you're up for it. Otherwise, perhaps punting it to @avalenzu?

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBWjPrq2OTYrnnnSoQV:b-v10aw2)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review pass done. PTAL.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[multibody/collision/test/collision_filter_group_test.cc, line 778 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBb79GJ3IOXaChgCqV9:-LBb79GJ3IOXaChgCqVA:b11bo7v) ([raw file](https://github.com/robotlocomotion/drake/blob/0b6a585a4bf62e611dd7fd8a1867f145ae20df94/multibody/collision/test/collision_filter_group_test.cc#L778)):*
> ```cpp
>   Element& element1b = **(body1.collision_elements_begin() + 1);
>   ASSERT_EQ(body2.get_num_collision_elements(), 0);
>   ASSERT_EQ(element1a.collision_cliques()[0], element1b.collision_cliques()[0]);
> ```

Should there be
```c++
  ASSERT_EQ(element1a.get_num_cliques(), 1);
  ASSERT_EQ(element1b.get_num_cliques(), 1);
```
before this line?

---

*[multibody/collision/test/collision_filter_group_test.cc, line 819 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBb7lUE4iQlz1P28hcC:-LBb7lUE4iQlz1P28hcD:bf6esij) ([raw file](https://github.com/robotlocomotion/drake/blob/0b6a585a4bf62e611dd7fd8a1867f145ae20df94/multibody/collision/test/collision_filter_group_test.cc#L819)):*
> ```cpp
>   ASSERT_EQ(body2.get_num_collision_elements(), 1);
>   Element& element2a = **body2.collision_elements_begin();
>   ASSERT_EQ(element1a.collision_cliques()[0], element2a.collision_cliques()[0]);
> ```

Do you want to check `get_num_cliques()` before this line?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBb8FyJCAfTT5_JcD2A:b837gv3)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Thanks @avalenzu . Comments responded. PTAL (Amazing you slogged through the tests to the fidelity you did.)

---

Review status: 10 of 13 files reviewed at latest revision, 17 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 482 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXJ4ll-aEYsX-ginWi:-LBb61ER7kH5S_lZSgec:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L482)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Unless I'm missing something, the notification doesn't occur until later when `NotifyFilterCriteriaChanged()` is called, correct? If so, could that part of the comment be moved?
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 625 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXKLFP53sW700gqrlY:-LBb6BO-5sWWDKEXp5rL:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L625)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Here you talk about bodies `P` and `Q`, but below it's `B` and `O`. Should these be consistent?
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 633 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXMqwX21YjXGaA7Y5W:-LBb6DTW60_Urfw6bxnJ:b8syn1o) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L633)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Since you've already ruled out the case where `test_body == body` with the previous `if`, could you use [`adjacentTo()`](https://github.com/RobotLocomotion/drake/blob/2b0e32a3d9c3e003dbe079cd2eccd640b43a94d0/multibody/rigid_body.h#L314) here instead? I think that would be much more intuitive given the fact that  the comments are all in terms of adjacency.
</blockquote></details>

Because the documentation says that cliques are applied to bodies if this method returns false. It means that this logic doesn't have to know *why* it returns false.

---

*[multibody/rigid_body_tree.cc, line 640 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXKfYy-Dau3WUlQo9d:-LBb6g7y2_hUf5G25kBy:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L640)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, I really like the shading! Could you align the `0`s though?
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 640 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXL8A_2W5IVmz325Ol:-LBb7FdC0qzqkyK12HwW:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L640)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, Making this "Queued" instead of "Pending" would tie it in with subscripts better.
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 669 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXODMS3Dj2p4oY_xHE:-LBb7LE_2QaFNCOReReE:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L669)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, missing period.
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 677 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXOaup7AzBG6Qld1rr:-LBb7_bcAuPp484cHorc:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L677)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Shouldn't this be something like "Add elements in B_Q to clique" instead?  Same comment for `O_Q`, `O_C`, and `B_C` below.
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 726 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXQKhL232LVgP3daCd:-LBb7gpJDJguyPedK_Qi:b-k1zkre) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L726)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, this comment appears to go with the previous if statement.
</blockquote></details>

It does. If it were a statement of what is being tested, I'd put it before the test. However, it's a statement of the state of the block, so I put it as the first thing inside the block.

---

*[multibody/collision/collision_filter_doxygen.h, line 288 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXRtE2CE6xeQzlvixa:-LBb8QwQ9pJ3FKTUgt-s:bhb20op) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/collision_filter_doxygen.h#L288)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> In addition, appending new bodies with collision elements will also automatically pick up the appropriate cliques to prevent self-collisions on a single body or collision between adjacent bodies.

It's unclear to me what the "new bodies" are being appended to here.
</blockquote></details>

Reworked this paragraph to address *this* comment and the next.

---

*[multibody/collision/collision_filter_doxygen.h, line 288 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXSWDQ21pTBC0muWtv:-LBb9HKJ0-JvrKt1kICy:bd4qvfe) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/collision_filter_doxygen.h#L288)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

This paragraph doesn' t mention adding new collision geometries to previously compiled bodies (unless I'm missing something).
</blockquote></details>

Part of the overhaul.

---

*[multibody/collision/test/collision_filter_group_test.cc, line 54 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXTNrrA61C2mCpE9yw:-LBb9LwQ1aNtAKkLaN-K:b-zcjvte) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L54)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, nice!
</blockquote></details>

Thanks

---

*[multibody/collision/test/collision_filter_group_test.cc, line 383 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXTjp73qPak-OD55JA:-LBb9rz4106E_KAKUtxq:b-886gsx) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L383)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

The comments talk about a floating joint, but this is a revolute joint. Why?
</blockquote></details>

Fixed

---

*[multibody/collision/test/collision_filter_group_test.cc, line 489 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXU4nY0faPcjNHt2Gw:-LBbF-qm-UwlOcV0UnZK:b32vhw) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L489)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

BTW, yay!
</blockquote></details>

:)

---

*[multibody/collision/test/collision_filter_group_test.cc, line 594 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXU_cd4YmUbh6RLIxA:-LBbA1hi8UARXRX2kYum:b44mj61) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L594)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Should "has collision elements" be "had previous collision elements" here? Same comment in subsequent states.
</blockquote></details>

There are two things in your comment:

Tense: I actively chose the present tense to describe all of these.  I.e., what is the state? It's this. Now I do the action. What is the result? It's now this. It's only the *current* state that determines the result for a given action.

Inclusion of the word "previous". I find this point compelling, and I'm going to be more clear about these. I've reworked all the language to be more explicit, complete, and consistent.

---

*[multibody/collision/test/collision_filter_group_test.cc, line 603 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXUssf6o1yX-iIwRwK:-LBbEa_p8Gpy70SZeCky:bu6xv6u) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L603)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Should "original elements" be "original element's cliques" here?
</blockquote></details>

See previous comment.

---

*[multibody/collision/test/collision_filter_group_test.cc, line 778 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBb79GJ3IOXaChgCqV9:-LBbEp3O7V6_aXTy18Ag:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/0b6a585a4bf62e611dd7fd8a1867f145ae20df94/multibody/collision/test/collision_filter_group_test.cc#L778)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Should there be
```c++
  ASSERT_EQ(element1a.get_num_cliques(), 1);
  ASSERT_EQ(element1b.get_num_cliques(), 1);
```
before this line?
</blockquote></details>

Done

---

*[multibody/collision/test/collision_filter_group_test.cc, line 819 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBb7lUE4iQlz1P28hcC:-LBbEy0k7AgIpMnDZEyJ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/0b6a585a4bf62e611dd7fd8a1867f145ae20df94/multibody/collision/test/collision_filter_group_test.cc#L819)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Do you want to check `get_num_cliques()` before this line?
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBb5b4P4z3Rd6_jD1uU:b-36gwwm)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
And back to you.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 726 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXQKhL232LVgP3daCd:-LBbNC3U1UiVduQ82aUU:bm8zbiq) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/rigid_body_tree.cc#L726)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Sorry, I meant that I would expect it to go with `if (element.get_num_cliques() == 0)`
</blockquote></details>

Ahhh...I see. I've modified the comment so it makes more sense in its present location.

---

*[multibody/collision/collision_filter_doxygen.h, line 282 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBbJ0Ch5LLq4K4F51Yw:-LBbNXhz3RrIVrTowbmi:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/21e17bc9c7c6171a94d7bed3a37c6886f7111aef/multibody/collision/collision_filter_doxygen.h#L282)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Incomplete sentence. Perhaps "These include:"?
</blockquote></details>

Done

---

*[multibody/collision/collision_filter_doxygen.h, line 287 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBbJEpF-utTtxoPPWmo:-LBbNebv1p7lNv47jQtv:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/21e17bc9c7c6171a94d7bed3a37c6886f7111aef/multibody/collision/collision_filter_doxygen.h#L287)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> ```
> If both bodies have joints
> ```
 How could they not have joints?
</blockquote></details>

Done

---

*[multibody/collision/test/collision_filter_group_test.cc, line 594 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBXU_cd4YmUbh6RLIxA:-LBbNrvZ6FTxSSP6YlnG:b-z8cs5x) ([raw file](https://github.com/robotlocomotion/drake/blob/7b2d609a20f94fb574d033da232cc58bda1a8ba0/multibody/collision/test/collision_filter_group_test.cc#L594)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Sounds good. States 1a,b,c are still in past tense, though.
</blockquote></details>

Touche

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBbNtfyDTnFKZGsxN4l:b-41x7g7)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBbaSAg1cIMDAXV-5aC:b-uwkowd)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+@soonho-tri for platform review, please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBbaz-jA_4yvQpPuq75:bxwr5c8)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
-@soonho-tri +@sherm1  I don't think @soonho-tri had a chance to get to this yesterday, so I'm rolling it to Friday's platform reviewer.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBfU4Hn4sWSfAMBAua6:b-gp9hqd)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
(unintended checkpoint)

Well.. I am still working on it. Please allow me more time. Sorry for the delay.

---

Reviewed 4 of 12 files at r1, 1 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/collision/model.h, line 81 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBfUAnM4a0VuBpnrHnn:-LBfUAnM4a0VuBpnrHno:b-wj36u) ([raw file](https://github.com/robotlocomotion/drake/blob/c400ae6844980e82904265ada1a2e558ef58d107/multibody/collision/model.h#L81)):*
> ```objc
>    * @param id  The id of the element whose filter criteria may have changed.
>    */
>   virtual void NotifyFilterCriteriaChanged(ElementId) {}
> ```

BTW, I'm wondering if it's better to make this pure virtual (`= 0`) and explicitly provide an empty body for `FclModel` and `UnusableModel` (It depends on what we expect for the new model class inheriting this one).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBfUxtE4Ux9QjW-vE1J:bl2yp6n)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Oops. Sorry, I just assumed that I'd hit you with it pretty late in the day. I'll undo that.

+@soonho-tri -@sherm1

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBfVd8-D9JGTzZbEDKV:brkmynt)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 2 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Pass completed. PTAL.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBg1f5L4Y78o78O1Img:bjv41yk)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Comments addressed, I believe.  PTAL

---

Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 484 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBfbaB1-kKMY4q4_Pjh:-LBv79Fd9gB7ZBk855Bb:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/c400ae6844980e82904265ada1a2e558ef58d107/multibody/rigid_body_tree.cc#L484)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

1) Can you check if we need to have `RigidBody<T>` instead of `RigidBody<double>` in this function body?
2) FYI, `std::pair<const RigidBody<double>*,` -> `std::pair<const RigidBody<double>* const,`
</blockquote></details>

Done

---

*[multibody/rigid_body_tree.cc, line 489 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBfdTaw6St1ne34cLaz:-LBv7Cz7A51hMYEXRkgg:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/c400ae6844980e82904265ada1a2e558ef58d107/multibody/rigid_body_tree.cc#L489)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Is there a way to rewrite this code without `const_cast`?

The phrase `simplifies the process` sounds like the answer is yes. If so I'd like to have a version without `const_cast`.

But I think the answer is actually negative (unless we change the type of `body_groups()` to be an unordered map from `RigidBody*` (instead of `const RigidBody*`) to `pair of bitmasks`.
</blockquote></details>

Done

---

*[multibody/collision/collision_filter.h, line 185 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBfVVlZDs8d1pMsYKYY:-LBv7LtrEimQNYZzyI1Z:byswykj) ([raw file](https://github.com/robotlocomotion/drake/blob/c400ae6844980e82904265ada1a2e558ef58d107/multibody/collision/collision_filter.h#L185)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, `specified specified` is over-specified.
</blockquote></details>

Done

Brought to you by the Redundant Department of Repetition and Redundancy.

---

*[multibody/collision/model.h, line 81 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8740#-LBfUAnM4a0VuBpnrHnn:-LBv7Xs92Bm4OEJPBTSL:bv2mybu) ([raw file](https://github.com/robotlocomotion/drake/blob/c400ae6844980e82904265ada1a2e558ef58d107/multibody/collision/model.h#L81)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I'm wondering if it's better to make this pure virtual (`= 0`) and explicitly provide an empty body for `FclModel` and `UnusableModel` (It depends on what we expect for the new model class inheriting this one).
</blockquote></details>

BTW I'm not overly concerned. These models are a dead end. `SceneGraph` doesn't make use of these. So, I *could* go strictly virtual and add empty implementations. But, because this feature only exists to accommodate `bullet`, I figured a coarser hammer would be appropriate.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8740#-:-LBvATsI-1Zf378OmIY_:b-ubsr28)*
<!-- Sent from Reviewable.io -->
